### PR TITLE
[WIP] Copy Comment Trivia from If Statement to Throw Expression

### LINKED ISSUE #38102: IDE0016 loses comments
**Version Used**: 
3.3.0-beta3-19413-06+ac06df1bffb7b7fd0e5bf63cc91921af76b21e03

**Steps to Reproduce**:
Any of these variants:
```C#
using System;

class Program
{
    static void Main() { }

    private object _arg;

    public Program(object arg)
    {
        if (arg == null)
        {
            throw new ArgumentNullException(nameof(arg)); // Oh no!
        }
        _arg = arg;
    }
}
```
or
```C#
using System;

class Program
{
    static void Main() { }

    private object _arg;

    public Program(object arg)
    {
        if (arg == null)
        {
            // Oh no!
            throw new ArgumentNullException(nameof(arg));
        }
        _arg = arg;
    }
}
```
or
```C#
using System;

class Program
{
    static void Main() { }

    private object _arg;

    public Program(object arg)
    {
        if (arg == null) // Oh no!
        {
            throw new ArgumentNullException(nameof(arg));
        }
        _arg = arg;
    }
}
```
Then apply IDE0016.

**Expected Behavior**:
```C#
using System;

class Program
{
    static void Main() { }

    private object _arg;

    public Program(object arg)
    {
        _arg = arg ?? throw new ArgumentNullException(nameof(arg)); // Oh no!
    }
}
```
or something similar, retaining the comment.  Or just not offering the fix.

**Actual Behavior**:
```C#
using System;

class Program
{
    static void Main() { }

    private object _arg;

    public Program(object arg)
    {
        _arg = arg ?? throw new ArgumentNullException(nameof(arg));
    }
}
```

### PR DESCRIPTION
**Pull request by 06needhamt**
Fixes #38102 

WIP because of an issue with the comment trivia being placed before the semicolon token pushing it onto the next line.

### CODE REVIEW

File: src/Features/Core/Portable/Microsoft.CodeAnalysis.Features.csproj [hunk 1]
```diff
@@ -16,6 +16,8 @@
   </PropertyGroup>
   <ItemGroup Label="Project References">
     <ProjectReference Include="..\..\..\Compilers\Core\Portable\Microsoft.CodeAnalysis.csproj" />
+    <ProjectReference Include="..\..\..\Compilers\CSharp\Portable\Microsoft.CodeAnalysis.CSharp.csproj" />
+    <ProjectReference Include="..\..\..\Compilers\VisualBasic\Portable\Microsoft.CodeAnalysis.VisualBasic.vbproj" />
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Microsoft.CodeAnalysis.Features.csproj]: remove.  this is language-neutral layer tht cannot reference language-specific layers :)

File: src/Features/Core/Portable/UseThrowExpression/UseThrowExpressionCodeFixProvider.cs [hunk 2]
```diff
@@ -1,13 +1,16 @@
...
+using System.Collections.Generic;
 using System.Collections.Immutable;
 using System.Composition;
 using System.Linq;
 using System.Threading;
 using System.Threading.Tasks;
 using Microsoft.CodeAnalysis.CodeActions;
 using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/UseThrowExpression/UseThrowExpressionCodeFixProvider.cs]: revert.  this layer cannot reference C#/VB specific code.

**Comment by CyrusNajmabadi** [comment]
> I’m still having an issue with the comments before the assignment as they are the trailing trivia of the open brace on the if statement

they shouldn't be.  can you show what you mean?

**Comment by 06needhamt** [comment]
@CyrusNajmabadi Upon further investigation it seems that the comments are the leading trivia from the throw keyword.

23E8B239-B17D-47A1-B4D0-9A50D191AE9B

**Comment by CyrusNajmabadi** [comment]
That makes sense :-). You should be able to pull the leading trivia for that statement and move it over

**Comment by 06needhamt** [comment]
@CyrusNajmabadi I am having trouble getting access to the `throw` keyword to capture the trivia. I already capture the throw statement trivia but that doesn't include the comment since it is the leading trivia of the throw keyword. Any idea of how I can capture it?

**Comment by CyrusNajmabadi** [comment]
Call GeyLeadingTrivia on the node :-)

**Comment by 06needhamt** [comment]
@CyrusNajmabadi I’ve managed to get all the cases working the only issue that’s left is the position of the semicolon but I’m not sure how to fix it.

**Comment by CyrusNajmabadi** [comment]
>  the only issue that’s left is the position of the semicolon but I’m not sure how to fix it.

Can you clarify that?

File: src/Features/Core/Portable/UseThrowExpression/UseThrowExpressionCodeFixProvider.cs [hunk 3]
```diff
@@ -54,13 +55,25 @@ public override Task RegisterCodeFixesAsync(CodeFixContext context)
                 var throwStatementExpression = root.FindNode(diagnostic.AdditionalLocations[1].SourceSpan);
                 var assignmentValue = root.FindNode(diagnostic.AdditionalLocations[2].SourceSpan);
 
-                // First, remote the if-statement entirely.
+                var text = document.GetTextSynchronously(cancellationToken);
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/UseThrowExpression/UseThrowExpressionCodeFixProvider.cs]: don't get this synchronously.  you're in an async conext.  so you can get it with the normal GetTextAsync.

File: src/Features/Core/Portable/UseThrowExpression/UseThrowExpressionCodeFixProvider.cs [hunk 4]
```diff
@@ -54,13 +55,25 @@ public override Task RegisterCodeFixesAsync(CodeFixContext context)
...
+                var throwLine = text.Lines.GetLineFromPosition(diagnostic.AdditionalLocations[1].SourceSpan.Start);
+
+                var fullThrowStatement = root.FindNode(throwLine.SpanIncludingLineBreak, getInnermostNodeForTie: true);
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/UseThrowExpression/UseThrowExpressionCodeFixProvider.cs]: don't do this.  you have the span of the node (from .AdditionalLocations).  Find that node, and then use that node to find the statement if it's there.

Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/UseThrowExpression/UseThrowExpressionCodeFixProvider.cs]: once you have the statement you can grab off the trivia as appropriate and move around.

File: src/Features/Core/Portable/UseThrowExpression/UseThrowExpressionCodeFixProvider.cs [hunk 5]
```diff
@@ -54,13 +55,25 @@ public override Task RegisterCodeFixesAsync(CodeFixContext context)
...
+                // First, remove the if-statement entirely.
                 editor.RemoveNode(ifStatement);
 
-                // Now, update the assignment value to go from 'a' to 'a ?? throw ...'.
-                editor.ReplaceNode(assignmentValue,
-                    generator.CoalesceExpression(assignmentValue,
-                    generator.ThrowExpression(throwStatementExpression)));
+                var newNode = generator.CoalesceExpression(assignmentValue,
+                    generator.ThrowExpression(throwStatementExpression))
+                    .WithTrailingTrivia(triviaList).NormalizeWhitespace();
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/UseThrowExpression/UseThrowExpressionCodeFixProvider.cs]: one problem is that you're adding the trailing trivia to the throw expression.  that's not correct (and is why the trivia is after the `)` but before the `;`).  You want to add the trailing trivia *after* the semicolon/statement that is being generated.

**Pull request closed without merging**